plume-macro/src/lib.rs

```diff
@@ -0,0 +1,139 @@
+#![recursion_limit = "128"]
+extern crate proc_macro;
+#[macro_use]
+extern crate quote;
+extern crate syn;
+
+use proc_macro::TokenStream;
+use proc_macro2::TokenStream as TokenStream2;
+use std::fs::{read_dir, File};
+use std::io::Read;
+use std::str::FromStr;
+
+#[proc_macro]
+pub fn import_migrations(input: TokenStream) -> TokenStream {
+    assert!(input.is_empty());
+    let migration_dir = if cfg!(feature = "postgres") {
+        "migrations/postgres"
+    } else if cfg!(feature = "sqlite") {
+        "migrations/sqlite"
+    } else {
+        "migrations"
+    };
+    let mut files = read_dir(migration_dir)
+        .unwrap()
+        .map(|dir| dir.unwrap())
+        .filter(|dir| dir.file_type().unwrap().is_dir())
+        .map(|dir| dir.path())
+        .collect::<Vec<_>>();
+    files.sort_unstable();
+    let migrations = files
+        .into_iter()
+        .map(|path| {
+            let mut up = path.clone();
+            let mut down = path.clone();
+            up.push("up.sql");
+            down.push("down.sql");
+            let mut up_sql = String::new();
+            let mut down_sql = String::new();
+            File::open(up).unwrap().read_to_string(&mut up_sql).unwrap();
+            File::open(down)
+                .unwrap()
+                .read_to_string(&mut down_sql)
+                .unwrap();
+            let name = path
+                .file_name()
+                .unwrap()
+                .to_str()
+                .unwrap()
+                .chars()
+                .filter(char::is_ascii_digit)
+                .take(14)
```

Comment:
this seems like a strange thing to `take()`
what is this whole expression supposed to do?

Reply:
diesel store whether migrations were run by storing a 14 digit code. For `2019-03-06-115158_blog_images`, it is translated to `20190306115158`. To do that I get an iterator over the folder name, filter only digits, and take the first 14 ones (if there are more, they are ignored)